File: 05_18_week12/EXERCISE/CPP/queue.cpp
#include <iostream>

#define MAX_QUEUE_SIZE 3        // queue의 최대 크기

typedef int element;            // 데이터 변수 정의
typedef struct {
    element queue[MAX_QUEUE_SIZE];
    int front, rear;
}Queue;                         // Queue 구조체

void init(Queue *q){            // Queue 초기화
    q->front = q->rear = -1;
}

bool is_full(Queue *q){         // Queue가 꽉 차 있는지 확인
    return q->rear == MAX_QUEUE_SIZE - 1;
}

bool is_empty(Queue *q){        // Queue가 비어있는지 확인
    return q->front == q->rear;
}

void push(Queue *q, element e){            // Queue 삽입
    if(is_full(q)){
        std::cerr << "Queue is full\n";
        exit(-1);
    }
    q->queue[++q->rear] = e;
}

element pop(Queue *q){          // Queue pop, (원소 삭제됨)
    if(is_empty(q)){
        std::cerr << "Queue is empty\n";
        exit(-1);
    }
    return q->queue[++q->front];
}

element peek(Queue *q){         // Queue peek, (원소 삭제되지않음)
    if(is_empty(q)){
        std::cerr << "Queue is empty\n";
        exit(-1);
    }
    return q->queue[q->front + 1];
}

int main(){
    Queue q;        // Queue 선언

    init(&q);       // queue초기화
    push(&q, 1);
    push(&q, 2);
    push(&q, 3);

    std::cout << std::boolalpha << is_full(&q) << "\n";     // true
    std::cout << pop(&q) << "\n";                           // 1
    std::cout << pop(&q) << "\n";                           // 2        
    std::cout << peek(&q) << "\n";                          // 3
    std::cout << pop(&q) << "\n";                           // 3
    std::cout << std::boolalpha << is_empty(&q) << "\n";    // true
}

File: 05_18_week12/EXERCISE/Python/Queue.py
MAX_QUEUE_SIZE = 3          # Queue의 최대 크기

class Queue:
    def __init__(self):
        self.queue = [0] * MAX_QUEUE_SIZE
        self.front = -1
        self.rear = -1
        
    def is_empty(self):         # Queue가 비어있는지 확인
        return self.front == self.rear
    
    def is_full(self):          # Queue가 꽉 차 있는지 확인
        return self.rear == MAX_QUEUE_SIZE - 1
    
    def push(self, data):       # Queue에 data 삽입
        if self.is_full():
            print("Queue is full")
            exit(-1)
        self.rear += 1
        self.queue[self.rear] = data
        
    def pop(self):               # Queue에서 data를 반환하고 삭제
        if self.is_empty():
            print("Queue is empty")
            exit(-1)
        return self.queue[self.front]
    
    def peek(self):              # Queue에서 data를 반환하고 삭제 X
        if self.is_empty():
            print("Queue is empty")
            exit(-1)
        return self.queue[self.front + 1]
    
if __name__ == "__main__":
    q = Queue()
    q.push(1)
    q.push(2)
    q.push(3)
    
    print(q.is_full())      # True
    print(q.pop())          # 1
    print(q.pop())          # 2
    print(q.peek())         # 3
    print(q.pop())          # 3
    print(q.is_empty())     # True